# Ex-9: Recommender Systems

## Question 1  (Recommendations by Critics) [20 points]

### Based on the example in Section 2.2 of Similarity and Recommender Systems:
### 1. [6 points] For a user named User3 who gave the following ratings:
            
            Australia: 6
            Hancock: 7
            Milk: 9
### What would the (Euclidean) distance be to each of the 6 critics?

In [ ]:
def EuclideanDist(user_rating, reviewer_rating):
    """
    Calculate the Euclidean distance between two rating vectors.
    :param user_rating: A list of ratings given by a user.
    :param reviewer_rating: A list of ratings given by a reviewer.
    :return: The Euclidean distance between the two rating vectors.
    """
    return sum((user_rating[i] - reviewer_rating[i]) ** 2 for i in range(len(user_rating))) ** 0.5

##User and Reviewer Ratings
user_rating = [6, 7, 9]
denby_scores = [3, 9, 9]
mccarthy_ratings = [7, 3, 8]
morgenstern_ratings = [7, 0, 8]
puig_ratings = [5, 5, 9]
travers_ratings = [5, 8, 10]
turan_ratings = [7, 4, 7]

# Calculate the Euclidean distance between the user rating and each reviewer's ratings
denby_distance = EuclideanDist(user_rating, denby_scores)
mccarthy_distance = EuclideanDist(user_rating, mccarthy_ratings)
morgenstern_distance = EuclideanDist(user_rating, morgenstern_ratings)
puig_distance = EuclideanDist(user_rating, puig_ratings)
travers_distance = EuclideanDist(user_rating, travers_ratings)
turan_distance = EuclideanDist(user_rating, turan_ratings)






In [6]:
print("Denby distance: ", denby_distance)
print("McCarthy distance: ", mccarthy_distance)
print("Morgenstern distance: ", morgenstern_distance)
print("Puig distance: ", puig_distance)
print("Travers distance: ", travers_distance)
print("Turan distance: ", turan_distance)

Denby distance:  3.605551275463989
McCarthy distance:  4.242640687119285
Morgenstern distance:  7.14142842854285
Puig distance:  2.23606797749979
Travers distance:  1.7320508075688772
Turan distance:  3.7416573867739413


### 2. [6 points] What is the similarity of User3 to each critic?

In [8]:
def similarity(user_rating, reviewer_rating):
    """
    Calculate the similarity between two rating vectors.
    :param user_rating: A list of ratings given by a user.
    :param reviewer_rating: A list of ratings given by a reviewer.
    :return: The similarity between the two rating vectors.
    """
    return 1 / (1 + EuclideanDist(user_rating, reviewer_rating))

In [10]:
similarity_denby = similarity(user_rating, denby_scores)
similarity_mccarthy = similarity(user_rating, mccarthy_ratings)
similarity_morgenstern = similarity(user_rating, morgenstern_ratings)
similarity_puig = similarity(user_rating, puig_ratings)
similarity_travers = similarity(user_rating, travers_ratings)
similarity_turan = similarity(user_rating, turan_ratings)
print("Denby similarity: ", similarity_denby)
print("McCarthy similarity: ", similarity_mccarthy)
print("Morgenstern similarity: ", similarity_morgenstern)   
print("Puig similarity: ", similarity_puig)
print("Travers similarity: ", similarity_travers)
print("Turan similarity: ", similarity_turan)

Denby similarity:  0.21712927295533244
McCarthy similarity:  0.1907435698305462
Morgenstern similarity:  0.122828568570857
Puig similarity:  0.3090169943749474
Travers similarity:  0.36602540378443865
Turan similarity:  0.21089672205953397


### 3. [4 points] Which critic is most similar to User3?

Morgenstern is the critic that is most similar to User3, with a similarity of 0.123

### 4. [4 points] Which film should be recommended to User3?

Using method 1 as outlined in the textbook, the recommendation would be Burn After Reading or Body of Lies, because they are tied for the highest rated film by Morenstern that User3 has not already reviewed. 

## Question 2 (The Ethics of Recommendation Systems) [80 points]

There are multiple ethical dimensions and potential issues with the use of recommendation systems. The research paper, “Recommender systems and their ethical challenges” and the textbook “Data Science in Context” discuss several of these issues. Three major issues that both discuss is the existence of inappropriate content, privacy implications, and bias. These are all very important issues with the use of recommendation systems that are intertwined with each other. 

The presence of offensive content being recommended by the system is a problem, whether that be religiously offensive, culturally offensive, or something else entirely. There are two main sides to this issue. The users of the recommendation system should not have to see offensive items being recommended to them when they are using a specific system. From a business perspective, this issue will likely cause customers to quit using the product and move onto other alternatives, ultimately causing the company some loss in profits. This ethical issue is unique in the sense that both sides want to avoid this issue. Customers do not want offensive content recommended and companies do not want their customers to have a poor experience that might cause them to seek out other products. 

However, the issue comes when trying to solve this universally accepted ethical dilemma with recommendation systems. The typical solution to this problem is that the company collects a significant amount of personal information about its customers to be able to curate their experience enough to avoid any offensive content being recommended. However, this leads to another ethical problem with these systems, privacy.  This collection of information, usually without explicit consent from the customers, is a very serious violation of privacy rights. This breach of privacy without proper consent is very dangerous, especially with the possibility of data leaks. Customers receive a benefit from a recommendation system that better understands them, but they also should be able to better control how their data is collected and processed. 

This discussion of privacy leads to another connected ethical concern with recommender systems, bias. When this amount of data is collected and processed, it can lead to biases and stereotypes being enforced within the recommendations themselves. A specific type of bias with recommendation systems discussed by the paper by Milano, Taddeo, and Floridi is news feeds. They discuss how the recommendations fed to the consumers can create a self-sustaining feedback loop. For example, a user that is relatively in the center politically can begin to get recommended conservative articles. After a long period of time, this could lead to the recommendation system altering this person’s personal beliefs. The loop becomes a feedback loop because as the user becomes more conservative, the system is increasingly encouraged to recommend conservative articles. This is especially true with the content recommendation of Facebook. According to an NPR article, “one out of five users experience an echo chamber on the platform, where at least three-quarters of the posts they see come from ideologically aligned sources” (NPR). This problem with bias becomes such a big issue when recommendation systems are allowed to run rampant without human intervention. 

Privacy, bias, and inappropriate content are just a few examples of the numerous ethical issues with recommendation systems and big data overall. Big Data has produced multiple innovations in our lives, but they also introduce serious ethical dilemmas. These ethical issues need to be seriously considered and addressed for the sake of our society. There have already been many negative events created or enforced by the presence of big data. It is important to begin to respond before it gets even worse. 

Source:
https://www.npr.org/2023/07/27/1190383104/new-study-shows-just-how-facebooks-algorithm-shapes-conservative-and-liberal-bub#:~:text=They%20found%20that%20on%20average,come%20from%20ideologically%20aligned%20sources. 
